Convert the `RefSet` primitive to a proper class and use a `Set` internally

timvandermeij opened the pull request:
The `RefSet` primitive predates ES6, so that most likely explains why an object is used internally to track the entries. However, nowadays we can use built-in JavaScript sets for this purpose. Built-in types are often more efficient/optimized and using it makes the code a bit more clear since we don't have to assign `true` to keys anymore just to indicate their presence.

timvandermeij commented:
/botio test

Snuffleupagus reviewed (changes requested):
Really sorry, but I'm not at all keen on these changes!

The `RefSet` abstraction is, as its name suggests, written specifically to operate only on `Ref` objects and I worry that simply using Sets directly in this manner makes the code more difficult to read/modify and that bugs are thus much easier to introduce.
Essentially, in the worker, it's generally preferable to avoid depending directly on the stringified version of a `Ref` object if possible.

Basically, unless there's test-cases where benchmarking shows *clear* improvements, then I'd ask that we please don't make these changes.

timvandermeij committed:
Convert the `RefSet` primitive to a proper class and use a `Set` internally

The `RefSet` primitive predates ES6, so that most likely explains why an
object is used internally to track the entries. However, nowadays we can
use built-in JavaScript sets for this purpose. Built-in types are often
more efficient/optimized and using it makes the code a bit more clear
since we don't have to assign `true` to keys anymore just to indicate
their presence.

timvandermeij commented:
I understand your concern here. I have updated the patch to keep `RefSet`, but instead refactor it to use a set internally instead of an object. Would that be better?

timvandermeij commented:
/botio unittest

Snuffleupagus commented:
> I have updated the patch to keep `RefSet`, but instead refactor it to use a set internally instead of an object. Would that be better?

Yes, that seems completely fine by me; thank you for doing this!

(Also, possibly this could make `RefSet.has()` more efficient since I recall some older issues where avoiding `in` usage helped performance.)

---

Just for good measure, let's run all tests :-)

/botio test

Snuffleupagus commented:
An idea could be to try and do something similar, but obviously with a `Map` instead, for the `RefSetCache`; I'd happily review such a patch :-)
